### ISSUE TITLE
SSH: Trying to connect to Tessel2 as root returns "Connection timed out" on port 22

### ISSUE THREAD

**Issue by vzamfirescu** [tessel/t2-cli#823]
Hello,

I tried connecting to my Tessel2 device today via ssh, but I keep getting the following error:
```
ssh -i ~/.tessel/id_rsa root@192.168.0.105
ssh: connect to host 192.168.0.105 port 22: Connection timed out
```

I tried running the ssh command with the ```ConnectTimeout``` flag set to higher values (20+ seconds), but I still get the same error.

I also deleted the ```~/.tessel/``` folder and ran ```t2 provision``` in order to refresh the keys used to authenticate on the device, but it did not fix my issue.

However, if I try running some code on the device, it runs with no issues whatsoever.

Furthermore, I've updated the code that I am running on the device via ```t2 run``` in order to write some data to an USB device mounted at ```/mnt/sda1/```, as I wanted to see if I am able to access the filesystem, and everything runs as expected.

After creating files with the ```fs``` module and writing data for a couple of minutes, I can then get the USB drive out of the device and check the data on my laptop.

Today is the first time I have encountered this strange behavior and I did no changes to my laptop in terms of firewall configuration, applications installed, npm globally installed packages etc.

As part of my investigation, I triggered a force update of the Tessel2 firmware via ```t2 update ---force```, but the errors keeps persisting.

After forcefully updating the firmware (even though I already had the latest version), I got the following error when I ran the first ```t2 list```:
```
INFO Searching for nearby Tessels...
ERR! Detected CLI crash { [Error: LIBUSB_TRANSFER_STALL] errno: 4 } Error: LIBUSB_TRANSFER_STALL
ERR!     at Error (native)
INFO Crash Reported: http://crash-reporter.tessel.io/crashes?fingerprint=-0x21dc75d3
```

However, it was a one-time only error that did not persist to any subsequent ```t2``` commands.

Below are some additional details regarding my setup:
**NODEJS**
```
node -v
v4.4.7

npm -v
2.15.8
```

**TESSEL2 DEVICE**
```
t2 provision
INFO Looking for your Tessel...
INFO Connected to beast.
INFO Authenticating Tessel with public key...
INFO Tessel is already authenticated with this computer.

t2 wifi
INFO Looking for your Tessel...
INFO Connected to <MY_TESSEL_DEVICE>.
INFO Connected to <MY_WIFI_SSID>
INFO IP Address: 192.168.0.105
INFO Signal Strength: (69/70)
INFO Bitrate: 15mbps
```

**Comment by johnnyman727** [comment]
Thanks for the report @vzamfirescu. Have you moved on to a different network since you last `ssh`'ed successfully?

**Comment by vzamfirescu** [comment]
@johnnyman727 
I didn't move to a different network than before.

Historically, I've only connected to 2 different wi-fi network when ```ssh-ing``` into my tessel device, and I usually provisioned the device all over again and everything worked like a charm.

**Comment by johnnyman727** [comment]
@vzamfirescu The only thing I can think of is that the service that runs SSH on T2 (I believe it's `dropbear`) is not running for whatever reason. Out of curiosity, have you power cycled the device since trying to connect? That would restart the `dropbear` daemon.

**Comment by vzamfirescu** [comment]
@johnnyman727  Yes, I shut it down several times and let it rest for more than several minutes before plugging it in again. The only thing that I have not tried yet is resetting it from the built-in button.

**Comment by johnnyman727** [comment]
I wouldn't worry about resetting it with the button - that doesn't do anything yet as far as I know.

The only other thing I can think of is that your computer and Tessel are on different networks. Besides that, I'll have to ponder for a bit.

**Comment by vzamfirescu** [comment]
I have a dynamic DNS set on my router, which makes it possible to access it remotely.

Would forwarding the ```22``` internal port on the IP address associated with my ```tessel``` and then trying to ```ssh``` into the device by using the dynamic DNS instead of the IP address help me out?

That way I can guarantee that I get access to my ```tessel```, regardless of them discovering each other, even though they are connected to the same ssid of the wi-fi network.

**Comment by johnnyman727** [comment]
@vzamfirescu I'm really not sure - I'm not much of an `ssh` or networking expert. I would suggest trying it and seeing what happens :)

**Comment by vzamfirescu** [comment]
I tried forwarding the tessel's internal port ```22``` to a DNS on port ```2222```, but I keep getting the same error.

Also, I tried running ```t2 root``` to log into the device, but I get the following error:
```
t2 root
INFO Looking for your Tessel...
WARN No Authorized Tessels Found.
```

However, when I run ```t2 provision```, this is the output that I get:
```
t2 provision
INFO Looking for your Tessel...
INFO Connected to <MY_TESSEL_DEVICE>.
INFO Authenticating Tessel with public key...
INFO Tessel is already authenticated with this computer.
```

**Comment by johnnyman727** [comment]
@vzamfirescu out of curiosity, when you run scripts on the device (deployed over USB, obviously) can those scripts send requests to the internet successfully?

**Comment by vzamfirescu** [comment]
I just ran the below snippet of code via ```t2 run```, and I keep getting ```getaddrinfo ENOTFOUND``` errors:
```
'use strict'
const fs = require('fs');
const http = require('http');

console.log('Code loaded on Tessel');

const options = {
    host: 'httpbin.org',
    path: '/ip',
    method: 'GET',
    headers: {
        'Content-Type': 'application/json'
    }
};

let data = '';

const req = http.request(options, (res) => {
    res.on('data', (chunk) => {
        console.log('Response Body: ', chunk);
    });
});

req.end();
req.on('error', (err) => {
    console.log('Error: ', err);
});
```

It appears that my tessel device doesn't have access to the Internet, even though when I run the above code via ```node```, I do get the proper response (a Buffer object).

How is it possible for my device not to have access to the Internet, even though all the other devices that are connected to the same router do?

Moreover, when I run ```t2 wifi```, this is the output that I get:
```
$ t2 wifi
INFO Looking for your Tessel...
INFO Connected to <MY_TESSEL_DEVICE>.
INFO Connected to <MY_WIFI_SSID>
INFO IP Address: 192.168.0.105
INFO Signal Strength: (70/70)
INFO Bitrate: 122mbps
```

Is there a way to reset the wifi adapter on the device? Or is there a possibility for other modules/components to interfere with this?

**Comment by johnnyman727** [comment]
Yeah, that's pretty interesting. I don't think any other modules/components would interfere.

Do you have a serial emulator like `screen` or `dterm`? If you do, you can get shell access to T2 and print out its wireless config settings. Something like `screen /dev/tty.usbmodem1422` and then once you're in the shell `cat /etc/config/wireless` and `cat /etc/config/network`. Those would be useful for debugging. Note that your device path is different depending on your OS and usb port.

**Comment by vzamfirescu** [comment]
@johnnyman727 

I managed to ```ssh``` into my tessel device by creating an access point with limited connectivity (no Internet access) - below is the output of the network configuration:
```
**cat /etc/config/network**
config interface 'loopback'
    option ifname 'lo'
    option proto 'static'
    option ipaddr '127.0.0.1'
    option netmask '255.0.0.0'

config interface 'wifi'
    option proto 'dhcp'

config interface 'ethernet'
    option ifname 'eth0'
    option proto 'dhcp'

config interface 'lan'
    option ifname 'wlan0'
    option proto 'static'
    option ipaddr '192.168.1.101'
    option netmask '255.255.255.0'

config switch 'mt7620'
    option enable_vlan 0
```

```
**cat /etc/config/wireless**

config wifi-device 'radio0'
        option type 'mac80211'
        option channel '11'
        option hwmode '11g'
        option path '10180000.wmac'
        option htmode 'HT20'
        option disabled '0'

config wifi-iface
        option device 'radio0'
        option network 'wifi'
        option mode 'sta'
        option ssid 'WIFI_2.4'
        option key '12345678'
        option encryption 'psk2'
        option disabled '0'

config wifi-iface
        option device 'radio0'
        option network 'lan'
        option mode 'ap'
        option ssid 'tessel_ap'
        option key '123456789'
        option encryption 'psk2'
        option disabled '0'
```

As previously mentioned, this is the content of those files when I connect from my laptop to the tessel access point created via ```t2 ap```, because otherwise I have no way of connecting to the device when it connects to my Archer C2 router.

**Comment by johnnyman727** [comment]
@vzamfirescu your `/etc/config/wireless` and `/etc/config/network` files look fine. I'm afraid I'm not sure how else to debug. @HipsterBrown might be able to suggest something.

**Comment by vzamfirescu** [comment]
@johnnyman727 Is there a way to reset my tessel device to factory settings? No matter how many times I force update it, uninstall the drivers, enter it in DFU mode, reinstall the drivers, the wifi doesn't work anymore.

I would like to at least be able to connect to wifi from a mobile hotspot until I find a fix for connecting it to my router.

Previously, even if it didn't work on my router, I could at least connect it to a mobile hotspot and ```t2 root``` into it.

**Comment by rwaldron** [comment]
`t2 restore`

:)

**Comment by johnnyman727** [comment]
I might actually suggest `t2 update -n` before `t2 restore` in this case. `t2 restore` will reflash the image used during manufacturing (it's pretty old) while `t2 update -n` will update with the most recent image but doesn't save any changed configurations.

**Comment by vzamfirescu** [comment]
@rwaldron @johnnyman727 

Thanks a lot for the info! :)

I ran ```t2 update -n``` and got the following output:
```
INFO Looking for your Tessel...
INFO Connected to <MY_TESSEL_DEVICE>.
WARN <MY_TESSEL_DEVICE> is already on the latest firmware version (0.0.13). You can force an update with "t2 update --force".
```

Afterwards, I ran ```t2 restore``` and successfully restored it to factory settings, after which I ran ```t2 update``` to update the firmware on my tessel device.

Now it seems that I can connect to my router via wi-fi, but I keep getting the following output, no matter how many times I run ```t2 provision```:
```
t2 list
INFO Searching for nearby Tessels...
        USB     Tessel-02A349B6A453
        LAN     Tessel-02A349B6A453     (USB connect and run `t2 provision` to authorize)

```

The only way I can now root into my tessel device is to connect both my laptop and the tessel to my mobile hotspot, and then the above ```t2 list``` error goes away after I run ```t2 provision```:
```
t2 list
INFO Searching for nearby Tessels...
        USB     Tessel-02A349B6A453
        LAN     Tessel-02A349B6A453
```

I literally tried everything from configuring my router in any possible way, isolating channels, setting a fixed channel width, allocating an IP address etc. and I am starting to believe that there is simply an incompatibility between the TP-LINK Archer C2 and the wifi module on the tessel.

I would be really interested in finding out if someone else with the same router encountered the same issue.

Until it will eventually get fixed, I guess I'll have to stick to mobile hotspots :)

**Comment by HipsterBrown** [comment]
@vzamfirescu For quick, wireless connection to the Tessel, you can start an access point using `t2 ap`, http://tessel.github.io/t2-start/ap.html, and connect your computer to that network, then you _should_ be able to provision and use `t2 root` from there. That's another possible compromise until we can fix that discovery issue.

**Comment by vzamfirescu** [comment]
Indeed, @HipsterBrown, thanks a lot for the tip!

The downside of that approach is that I won't be able to run any code that involves a web server or any kind of HTTP requests.

**Comment by HipsterBrown** [comment]
@vzamfirescu If Tessel is connected to a wifi network as well, then the HTTP requests _should_ still work. But your computer will be disconnected from an Internet-connected network, yes.

**Comment by rwaldron** [comment]
> it's pretty old


Every new Tessel 2 I've received has 0.0.13, out of the box

**Comment by vzamfirescu** [comment]
@rwaldron The ```t2 restore``` command reverted the firmware on my device to 0.0.12

@HipsterBrown That is the issue that I've tried debugging for the past couple of days - even though my tessel device connects to my wi-fi network, it gets assigned an IP, but has no Internet access.

I tried pinging a website via the ```http``` module, but no response.

The only way that I can currently get Internet access on my tessel device is to connect it to a mobile hotspot.

Somehow, it appears there is a conflict between my TP-LINK Archer C2 router and the wi-fi module on the tessel, even though previously it worked just fine.

I have also previously tested the wi-fi by connecting to an ASUS RT-N53 router with the same ISP and everything worked flawlessly.

As far as I can understand from my previous debugging on the Archer C2, it appears that the error might be with ```t2 provision```, as no matter how many times I run it, I get the below output even if my tessel successfully connects to the router and gets assigned an IP address:
```
t2 list
INFO Searching for nearby Tessels...
        USB     Tessel-02A349B6A453
        LAN     Tessel-02A349B6A453     (USB connect and run `t2 provision` to authorize)
```
